packages/logging/src/analytics-helpers.ts

```diff
@@ -119,4 +159,206 @@ export class ToggleableAnalytics implements MongoshAnalytics {
+  let intervalId: ReturnType<typeof setInterval>;
+  const lockfilePath = `${filepath}.lock`;
+  const unlock = async() => {
+    clearInterval(intervalId);
+    try {
+      return await fs.promises.rmdir(lockfilePath);
+    } catch {
+      // ignore update errors
+    }
+  };
+  try {
+    await fs.promises.mkdir(lockfilePath);
+    // Set up an interval update for lockfile mtime so that if the lockfile is
+    // created by long running process (longer than staleDuration) we make sure
+    // that another process doesn't consider lockfile stale
+    intervalId = setInterval(() => {
+      const now = Date.now();
+      fs.promises.utimes(lockfilePath, now, now).catch(() => {});
+    }, staleDuration / 2);
+    // eslint-disable-next-line chai-friendly/no-unused-expressions
+    intervalId.unref?.();
+    return unlock;
+  } catch (e) {
+    if ((e as any).code !== 'EEXIST') {
+      throw e;
+    }
+    const stats = await fs.promises.stat(lockfilePath);
+    // To make sure that the lockfile is not just a leftover from an unclean
+    // process exit, we check whether or not it is stale
+    if (Date.now() - stats.mtimeMs > staleDuration) {
+      await fs.promises.rmdir(lockfilePath);
+      return lockfile(filepath, staleDuration);
+    }
+    throw new Error(`File ${filepath} already locked`);
+  }
+}
+
+export class ThrottledAnalytics implements MongoshAnalytics {
+  private trackQueue = new Queue<AnalyticsTrackMessage>((message) => {
+    if (this.shouldEmitAnalyticsEvent()) {
+      this.target.track(message);
+      this.throttleState.count++;
+    }
+  });
+  private target: ThrottledAnalyticsOptions['target'] = new NoopAnalytics();
+  private currentUserId: string | null = null;
+  private throttleOptions: ThrottledAnalyticsOptions['throttle'] = null;
+  private throttleState = { count: 0, timestamp: Date.now() };
+  private restorePromise: Promise<void> = Promise.resolve();
+  private unlock: () => Promise<void> = () => Promise.resolve();
+
+  constructor({ target, throttle }: Partial<ThrottledAnalyticsOptions> = {}) {
+    this.target = target ?? this.target;
+    this.throttleOptions = throttle ?? this.throttleOptions;
+  }
+
+  get metadataPath() {
+    if (!this.throttleOptions) {
+      throw new Error(
+        'Metadata path is not avaialble if throttling is disabled'
+      );
+    }
+
+    if (!this.currentUserId) {
+      throw new Error('Metadata path is not avaialble if userId is not set');
+    }
+
+    const {
+      throttleOptions: { metadataPath },
+      currentUserId: userId
+    } = this;
+
+    return path.resolve(metadataPath, `am-${userId}.json`);
+  }
+
+  identify(message: AnalyticsIdentifyMessage): void {
+    if (this.currentUserId) {
+      throw new Error('Identify can only be called once per user session');
+    }
+    this.currentUserId = message.userId ?? message.anonymousId;
+    console.log(this.currentUserId);
+    this.restorePromise = this.restoreThrottleState().then((enabled) => {
```

Comment:
What do we do when `restoreThrottleState()` rejects, e.g. because `readFile()` fails? Right now, we'd get an unhandled rejection here.

Reply:
My assumption was that `shellHomeDirectory.ensureExists` would make sure that these are available or will disable telemetry so we never hit this. It is fair to expect something still going wrong though and so we probably don't want this to break shell as this is telemetry? I'll change `restoreThrottleState` to return `false` if something happened when reading a file other than ENOENT

Reply:
Yeah, I’d definitely prefer to do something else with the error instead of an unhandled rejection, and bubbling it up to the user doesn’t seem quite right since as you mention, this is for telemetry only. Could also just be logging it instead